### PR TITLE
Adds extra safety check for csi GC

### PR DESCRIPTION
**Pull request by 0sewa0**
# Description

It can happen that the GC runs while a Dynakube is still downloading the codemodules image, and the GC just deletes the directory because it doesn't know according to the database if that codemodules image is used or not.
This is because we only set the digest/latestVersion in the database when said download is finished.

## How can this be tested?
Deploy >1 Dynakube at the same time, using the same `codeModulesImage`.

## Checklist
- [x] Unit tests have been updated/added
- [x] PR is labeled accordingly
- [x] I have read and understood the [contribution guidelines](/CONTRIBUTING.md)

### CODE REVIEW

File: src/controllers/csi/gc/controller.go [hunk 1]
```diff
@@ -132,17 +142,31 @@ func collectGCInfo(ctx context.Context, apiReader client.Reader, dynakube dynatr
 	}, nil
 }
 
+func isSafeToGC(access metadata.Access, dynakubeList *dynatracev1beta1.DynaKubeList) bool {
+	dkMetadataList, err := access.GetAllDynakubes()
+	if err != nil {
+		log.Info("failed to get dynakube metadata from database, err: %s")
```
Comment (mjgrzybek) [inline src/controllers/csi/gc/controller.go:148]:
```suggestion
		log.Info("failed to get dynakube metadata from database", "err", err)
```

**Review by mjgrzybek — CHANGES_REQUESTED** [review]

File: src/controllers/csi/gc/controller.go [hunk 2]
```diff
@@ -132,17 +142,31 @@ func collectGCInfo(ctx context.Context, apiReader client.Reader, dynakube dynatr
...
+		return false
+	}
+	filteredDynakubes := filterCodeModulesImageDynakubes(dynakubeList)
+	for _, dkMetadata := range dkMetadataList {
+		if dkMetadata.LatestVersion == "" {
+			return false
+		}
+		dynakube, ok := filteredDynakubes[dkMetadata.Name]
```
Comment (meik99) [inline src/controllers/csi/gc/controller.go:156]:
```suggestion
		dynakube, hasDynakube := filteredDynakubes[dkMetadata.Name]
```

File: src/controllers/csi/gc/controller.go [hunk 3]
```diff
@@ -132,17 +142,31 @@ func collectGCInfo(ctx context.Context, apiReader client.Reader, dynakube dynatr
...
 
+func isSafeToGC(access metadata.Access, dynakubeList *dynatracev1beta1.DynaKubeList) bool {
+	dkMetadataList, err := access.GetAllDynakubes()
+	if err != nil {
+		log.Info("failed to get dynakube metadata from database, err: %s")
+		return false
+	}
+	filteredDynakubes := filterCodeModulesImageDynakubes(dynakubeList)
+	for _, dkMetadata := range dkMetadataList {
+		if dkMetadata.LatestVersion == "" {
```
Comment (meik99) [inline src/controllers/csi/gc/controller.go:153]:
```suggestion
		if isDownloading(dkMetadata.LatestVersion) {
		
func isDownloading(latestVersion string) bool {
  return  latestVersion == ""
}
```

would make it clear why it is not safe to gc in that scenario

File: src/controllers/csi/gc/controller.go [hunk 4]
```diff
@@ -132,17 +142,31 @@ func collectGCInfo(ctx context.Context, apiReader client.Reader, dynakube dynatr
...
+			return false
+		}
+		dynakube, ok := filteredDynakubes[dkMetadata.Name]
+		if ok && dynakube.CodeModulesVersion() != dkMetadata.LatestVersion {
```
Comment (meik99) [inline src/controllers/csi/gc/controller.go:157]:
```suggestion
		if ok && isUpdating(dynakube.CodeModulesVersion(), dkMetadata.LatestVersion) {
		
func isUpdating(codeModulesVersion string, latestVersion string) bool {
  return dynakube.CodeModulesVersion() != dkMetadata.LatestVersion
}
```

File: src/controllers/csi/gc/controller.go [hunk 5]
```diff
@@ -152,7 +176,29 @@ func getAllPinnedVersionsForTenantUUID(ctx context.Context, apiReader client.Rea
...
-		pinnedImages[dynakube.CodeModulesVersion()] = true
+		codeModuleVersion := dynakube.CodeModulesVersion()
+		if codeModuleVersion != "" {
+			pinnedVersions[codeModuleVersion] = true
+		}
+	}
+	return pinnedVersions, nil
+}
+
+func getAllDynakubes(ctx context.Context, apiReader client.Reader, namespace string) (*dynatracev1beta1.DynaKubeList, error) {
```
Comment (meik99) [inline src/controllers/csi/gc/controller.go:187]: i might be wrong, but it feels like this function already exists.
maybe check if it does and refactor accordingly, or ignore this comment if it doesn't

**Review by meik99 — CHANGES_REQUESTED** [review]

File: src/controllers/csi/gc/controller.go (hunk 2 above)
Comment (mjgrzybek) [reply src/controllers/csi/gc/controller.go:156]: Imho `ok` is better because it's standard way of checking it item exists in a map.

Comment (0sewa0) [reply src/controllers/csi/gc/controller.go:156]: I moved this part into a separate (2 line) function and in that case `ok` is better IMO

File: src/controllers/csi/gc/controller.go (hunk 5 above)
Comment (0sewa0) [reply src/controllers/csi/gc/controller.go:195]: didn't find anything 😢

**Pull request merged**